apps/condo/domains/common/components/Prompt.tsx

```diff
@@ -21,51 +21,96 @@ const Prompt: React.FC<IPromptProps> = ({ children, title, form, handleSave: for
     const hideModal = () => setIsModalVisible(false)
-    const handleCancel = () => {
+    const handleCancel = async () => {
         isIgnoringPrompt.current = true
         hideModal()
-        router.push(next)
+
+        if (isBackButtonClicked.current) {
+            router.back()
+        }
+
+        await router.push(next)
     }
-    const handleSave = () => {
-        isIgnoringPrompt.current = true
+    const handleSave = async () => {
         hideModal()
-        formSubmit()
+        const isValid = await form.validateFields()
+
+        if (isValid) {
+            isIgnoringPrompt.current = true
+            formSubmit()
+
+            if (isBackButtonClicked.current) {
+                router.back()
+            }
+        }
     }
     // TODO(zuch): find a solution to watch for file changes and contact changes
     // as they are using custom hooks and their fields do not present on initial form state
     const isFormChanged = () => {
         const newFormFields = pick(form.getFieldsValue(), Object.keys(initialFormState.current))
         return !isEqual(initialFormState.current, newFormFields)
     }
+
     useEffect(() => {
-        initialFormState.current = form.getFieldsValue()
+        if (typeof window === 'undefined') {
+            return
+        }
+
         // Todo(zuch): find a better way to turn off Prompt on form submit
         const oldFormSubmit = form.submit
-        form.submit = () => {
-            isIgnoringPrompt.current = true
-            oldFormSubmit.call(form)
+        initialFormState.current = form.getFieldsValue()
+
+        form.submit = async () => {
+            const isValid = await form.validateFields()
+
+            if (isValid) {
+                isIgnoringPrompt.current = true
+                oldFormSubmit.call(form)
+            }
         }
+
         const onRouteChange = url => {
-            if (!isIgnoringPrompt.current) {
-                if (isFormChanged()) {
-                    setNext(url)
-                    showModal()
-                    // Todo(zuch): wait for next.js implement router abort method and remove custom error
-                    // https://github.com/vercel/next.js/issues/2476
-                    throw 'Preventing form from close (ignore this error)'
-                }
+            if (!isIgnoringPrompt.current && isFormChanged()) {
+                setNext(url)
+                showModal()
+
+                // Todo(zuch): wait for next.js implement router abort method and remove custom error
+                // https://github.com/vercel/next.js/issues/2476
+                throw 'Preventing form from close (ignore this error)'
+            }
+        }
+
+        const handleBeforePopState = ({ as }) => {
+            isBackButtonClicked.current = true
+
+            if (!isIgnoringPrompt.current && isFormChanged()) {
+                setNext(as)
+                showModal()
+
+                window.history.pushState(null, '', router.asPath)
```

Comment:
Why not a router from next?

Reply:
We need to manage history state, not a app route.
Clicking the back button in browser causes a `popState` event that changes the history state (but the next page does not change because of `onRouteChange` handler). To get history back, you must run `pushState`